File: sources/LiquidSense Release/me/aquavit/liquidsense/ui/client/hud/element/Element.java
package me.aquavit.liquidsense.ui.client.hud.element;

import me.aquavit.liquidsense.utils.mc.MinecraftInstance;
import me.aquavit.liquidsense.value.Value;
import net.minecraft.client.gui.ScaledResolution;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

public abstract class Element extends MinecraftInstance {

	private final ElementInfo info;
	public float prevMouseX;
	public float prevMouseY;
	private double x;
	private double y;
	private float scale;
	private Border border;
	private boolean drag;
	private Side side;

	public Element(double x, double y, float scale, Side side) {
		this.x = x;
		this.y = y;
		this.side = side;
		ElementInfo elementInfo = this.getClass().getAnnotation(ElementInfo.class);
		if (elementInfo != null) {
			this.info = elementInfo;
			this.scale = 1.0F;
			this.setScale(scale);
		} else {
			throw new IllegalArgumentException("Passed element with missing element info");
		}
	}

	public Element() {
		this(2.0, 2.0, 1F, new Side(Side.Horizontal.LEFT, Side.Vertical.UP));
	}

	public final ElementInfo getInfo() {
		return this.info;
	}

	public final float getScale() {
		return this.info.disableScale() ? 1.0F : this.scale;
	}

	public final void setScale(float value) {
		if (!this.info.disableScale()) {
			this.scale = value;
		}
	}

	public final String getName() {
		return this.info.name();
	}

	public final double getRenderX() {
		switch (side.getHorizontal()) {
			case LEFT:
				return x;
			case RIGHT:
				return new ScaledResolution(mc).getScaledWidth() - x;
			case MIDDLE:
				return (float) new ScaledResolution(mc).getScaledWidth() / 2 - x;
		}
		return 0;
	}

	public final double getRenderY() {
		switch (side.getVertical()) {
			case UP:
				return y;
			case MIDDLE:
				return (float) (new ScaledResolution(mc).getScaledHeight() / 2) - y;
			case DOWN:
				return new ScaledResolution(mc).getScaledHeight() - y;
			default:return 0;
		}
	}

	public final void setRenderX(double value) {
		switch (side.getHorizontal()) {
			case MIDDLE:
			case RIGHT:
				x -= value;
				break;
			case LEFT:
				x += value;
				break;
		}
	}

	public final void setRenderY(double value) {
		switch (side.getVertical()) {
			case UP: {
				y += value;
				break;
			}
			case MIDDLE:
			case DOWN:
				y -= value;
				break;
		}
	}

	public final Border getBorder() {
		return this.border;
	}

	public final void setBorder(Border state) {
		this.border = state;
	}

	public final boolean getDrag() {
		return this.drag;
	}

	public final void setDrag(boolean state) {
		this.drag = state;
	}

	public final float getPrevMouseX() {
		return this.prevMouseX;
	}

	public final void setPrevMouseX(float x) {
		this.prevMouseX = x;
	}

	public final float getPrevMouseY() {
		return this.prevMouseY;
	}

	public final void setPrevMouseY(float y) {
		this.prevMouseY = y;
	}

	public List<Value<?>> getValues() {
		final List<Value<?>> values = new ArrayList<Value<?>>();
		for (final Field field : this.getClass().getDeclaredFields()) {
			try {
				field.setAccessible(true);
				final Object o = field.get(this);
				if (o instanceof Value && ((Value<?>) o).isDisplayable()) {
					values.add((Value<?>) o);
				}
			} catch (IllegalAccessException e) {
				e.printStackTrace();
			}
		}
		return values;
	}

	public boolean createElement() {
		return true;
	}

	public void destroyElement() {
	}

	public abstract Border drawElement();

	public void updateElement() {
	}

	public void livingupdateElement() {
	}

	public boolean isInBorder(double x, double y) {
		Border border = this.border;
		if (border == null) return false;

		float minX = Math.min(border.x, border.x2);
		float minY = Math.min(border.y, border.y2);

		float maxX = Math.max(border.x, border.x2);
		float maxY = Math.max(border.y, border.y2);

		return minX <= x && minY <= y && maxX >= x && maxY >= y;
	}

	public void handleMouseClick(double x, double y, int mouseButton) {
	}

	public void handleKey(char c, int keyCode) {
	}

	public final double getX() {
		return this.x;
	}

	public final void setX(double x) {
		this.x = x;
	}

	public final double getY() {
		return this.y;
	}

	public final void setY(double y) {
		this.y = y;
	}

	public final Side getSide() {
		return this.side;
	}

	public final void setSide(Side side) {
		this.side = side;
	}
}

File: sources/Kevin Reborn/kevin/hud/element/Element.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.hud.element

import kevin.module.Value
import kevin.utils.MinecraftInstance
import kevin.utils.RenderUtils
import net.minecraft.client.gui.ScaledResolution
import kotlin.math.max
import kotlin.math.min

abstract class Element(var x: Double = 2.0, var y: Double = 2.0, scale: Float = 1F,
                       var side: Side = Side.default()) : MinecraftInstance() {
    val info = javaClass.getAnnotation(ElementInfo::class.java)
        ?: throw IllegalArgumentException("Passed element with missing element info")

    var scale: Float = 1F
        set(value) {
            if (info.disableScale)
                return

            field = value
        }
        get() {
            if (info.disableScale)
                return 1.0f
            return field
        }

    init {
        this.scale = scale
    }

    val name: String
        get() = info.name

    var renderX: Double
        get() = when (side.horizontal) {
            Side.Horizontal.LEFT -> x
            Side.Horizontal.MIDDLE -> (ScaledResolution(mc).scaledWidth / 2) - x
            Side.Horizontal.RIGHT -> ScaledResolution(mc).scaledWidth - x
        }
        set(value) = when (side.horizontal) {
            Side.Horizontal.LEFT -> {
                x += value
            }
            Side.Horizontal.MIDDLE, Side.Horizontal.RIGHT -> {
                x -= value
            }
        }

    var renderY: Double
        get() = when (side.vertical) {
            Side.Vertical.UP -> y
            Side.Vertical.MIDDLE -> (ScaledResolution(mc).scaledHeight / 2) - y
            Side.Vertical.DOWN -> ScaledResolution(mc).scaledHeight - y
        }
        set(value) = when (side.vertical) {
            Side.Vertical.UP -> {
                y += value
            }
            Side.Vertical.MIDDLE, Side.Vertical.DOWN -> {
                y -= value
            }
        }

    var border: Border? = null

    var drag = false
    var prevMouseX = 0F
    var prevMouseY = 0F

    /**
     * Get all values of element
     */
    open val values: List<Value<*>>
        get() = javaClass.declaredFields.map { valueField ->
            valueField.isAccessible = true
            valueField[this]
        }.filterIsInstance<Value<*>>()

    /**
     * Called when element created
     */
    open fun createElement() = true

    /**
     * Called when element destroyed
     */
    open fun destroyElement() {}

    /**
     * Draw element
     */
    abstract fun drawElement(): Border?

    /**
     * Update element
     */
    open fun updateElement() {}

    /**
     * Check if [x] and [y] is in element border
     */
    open fun isInBorder(x: Double, y: Double): Boolean {
        val border = border ?: return false

        val minX = min(border.x, border.x2)
        val minY = min(border.y, border.y2)

        val maxX = max(border.x, border.x2)
        val maxY = max(border.y, border.y2)

        return minX <= x && minY <= y && maxX >= x && maxY >= y
    }

    /**
     * Called when mouse clicked
     */
    open fun handleMouseClick(x: Double, y: Double, mouseButton: Int) {}

    /**
     * Called when key pressed
     */
    open fun handleKey(c: Char, keyCode: Int) {}

}

/**
 * Element info
 */
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
annotation class ElementInfo(val name: String, val single: Boolean = false, val force: Boolean = false, val disableScale: Boolean = false, val priority: Int = 0)

/**
 * CustomHUD Side
 *
 * Allows to change default x and y position by side
 */
class Side(var horizontal: Horizontal, var vertical: Vertical) {

    companion object {

        /**
         * Default element side
         */
        fun default() = Side(Horizontal.LEFT, Vertical.UP)

    }

    /**
     * Horizontal side
     */
    enum class Horizontal(val sideName: String) {

        LEFT("Left"),
        MIDDLE("Middle"),
        RIGHT("Right");

        companion object {

            @JvmStatic
            fun getByName(name: String) = values().find { it.sideName == name }

        }

    }

    /**
     * Vertical side
     */
    enum class Vertical(val sideName: String) {

        UP("Up"),
        MIDDLE("Middle"),
        DOWN("Down");

        companion object {

            @JvmStatic
            fun getByName(name: String) = values().find { it.sideName == name }

        }

    }

}

/**
 * Border of element
 */
data class Border(val x: Float, val y: Float, val x2: Float, val y2: Float) {

    fun draw() = RenderUtils.drawBorderedRect(x, y, x2, y2, 3F, Int.MIN_VALUE, 0)

}